--- da-ebola-v7-test/da-ebola-v7-test.dnaorg_annotate.sqtable	2018-09-11 13:45:58.923074000 -0400
+++ /panfs/pan1/infernal/notebook/18_0722_virus_duplicate_features/dnaorg_scripts/testfiles/expected-files/ebola.v7.noutr.nocdsgene.nocodonstart.sqtable	2018-09-11 13:47:16.322452000 -0400
@@ -7,0 +8 @@
+
@@ -37,0 +39,3 @@
+
+
+
@@ -44,0 +49 @@
+
@@ -74,0 +80,5 @@
+
+Additional note(s) to submitter:
+WARNING: Partial sequence: This sequence lacks 507 nucleotides of sequences at the 3' end
+
+
@@ -81,0 +92 @@
+
@@ -107,0 +119,6 @@
+
+Additional note(s) to submitter:
+WARNING: Partial sequence: This sequence lacks 304 nucleotides of coding sequences at the 3' end
+
+
+
@@ -119,2 +136,2 @@
-4479	5372	CDS
-			product	matrix protein
+4479	5372	misc_feature
+			note	similar to matrix protein
@@ -144,0 +162,5 @@
+
+Additional note(s) to submitter:
+ERROR: Deletion of Nucleotides: Contains internal deletion of >10 nucleotides in matrix protein
+
+
@@ -184 +206,4 @@
-ERROR: CDS Has Stop Codon: Contains unexpected stop codon in matrix protein; in-frame stop codon exists 5' of stop position predicted by homology to reference [homology search predicted 4479..5371 revised to 4479..5084 (stop shifted 287 nt)]
+ERROR: Deletion of Nucleotides: Contains internal deletion of >10 nucleotides in matrix protein
+
+
+
@@ -221,0 +247,6 @@
+
+Additional note(s) to submitter:
+WARNING: Partial sequence: Lacks 341 nucleotides from 5’ end
+
+
+
@@ -254,0 +286,6 @@
+
+Additional note(s) to submitter:
+WARNING: Partial sequence: Lacks 3569 nucleotides from 5’ end
+
+
+
@@ -294,0 +332,3 @@
+
+
+
@@ -332 +371,0 @@
-
@@ -334 +373,4 @@
-ERROR: CDS Has Stop Codon: Contains unexpected stop codon in RNA-dependent RNA polymerase; in-frame stop codon exists 5' of stop position predicted by homology to reference [homology search predicted 11581..18219 revised to 11581..16236 (stop shifted 1983 nt)]
+ERROR: CDS Has Stop Codon: Contains unexpected stop codon in RNA-dependent RNA polymerase
+
+
+
@@ -374,0 +417,2 @@
+
+
@@ -381,0 +426 @@
+
@@ -408,4 +453,4 @@
-18311	11523	misc_feature
-			note	similar to L
-18219	11581	misc_feature
-			note	similar to RNA-dependent RNA polymerase
+11501	18282	gene
+			gene	L
+18219	11581	CDS
+			product	RNA-dependent RNA polymerase
@@ -414,2 +459,3 @@
-ERROR: Reverse Complement: Sequence may be misassembled; RNA-dependent RNA polymerase appears to be reverse complemented
-ERROR: Reverse Complement: Sequence may be misassembled; L appears to be reverse complemented
+ERROR:  Reverse Complement: Sequence appears to be missassembled
+
+
@@ -422,0 +469 @@
+
@@ -455 +502,4 @@
-ERROR: CDS Has Stop Codon: Contains unexpected stop codon in RNA-dependent RNA polymerase; in-frame stop codon exists 5' of stop position predicted by homology to reference [homology search predicted 11581..18218 revised to 11581..12768 (stop shifted 5450 nt)]
+ERROR:  Deletion of Nucleotides: Contains internal deletion of 1 nucleotide(s) in RNA-dependent RNA polymerase
+ERROR: CDS Has Frameshift: Contains internal deletion of 1 nucleotide(s) in RNA-dependent RNA polymerase
+
+
@@ -462,0 +513 @@
+
@@ -492,0 +544,5 @@
+
+
+
+
+
@@ -525,0 +582,5 @@
+
+Additional note(s) to submitter:
+WARNING: Partial sequence: Lacks 3385 nucleotides from 5’ end
+
+
@@ -531 +592,4 @@
-			note	more than one homologous region [nucleoprotein: 2 regions]
+<3027	5583	gene
+			gene	NP
+3027	5246	CDS
+			product	nucleoprotein
@@ -533,0 +598 @@
+
@@ -566 +631,4 @@
-ERROR: Duplicate Feature: more than one homologous region [nucleoprotein: 2 regions]
+ERROR: Wrong ORF Number: Sequence appears to be missassembled
+
+
+
@@ -579,0 +648,5 @@
+
+Additional note(s) to submitter:
+WARNING: Partial sequence: Lacks 6123 nucleotides from 5’ end; Lacks 11509 nucleotides from 3' end
+
+
@@ -590,0 +664,5 @@
+
+Additional note(s) to submitter:
+WARNING: Partial sequence: Lacks 6330 nucleotides from 5’ end; Lacks 12325 nucleotides from 3' end
+
+
@@ -597,0 +676 @@
+
@@ -630 +709,8 @@
-ERROR: CDS Has Stop Codon: Contains unexpected stop codon in RNA-dependent RNA polymerase; in-frame stop codon exists 5' of stop position predicted by homology to reference [homology search predicted 11581..18185 revised to 11581..11889 (stop shifted 6296 nt)]
+WARNING: This sequence contains 72 non-ACTG character(s) in minor nucleoprotein
+WARNING: This sequence contains 11 non-ACTG character(s) in membrane-associated protein
+WARNING: This sequence contains 11 non-ACTG character(s) in RNA-dependent RNA polymerase
+ERROR: Deletion of Nucleotides: Contains internal deletion of ≥10nt nucleotide(s) in RNA-dependent RNA polymerase
+ERROR: CDS Has Frameshift: Contains internal deletion of ≥10nt nucleotide(s) in RNA-dependent RNA polymerase
+
+
+
@@ -667,0 +754,8 @@
+
+Additional note(s) to submitter:
+WARNING: This sequence contains 1 non-ACTG character(s) in matrix protein
+WARNING: This sequence contains 2 non-ACTG character(s) in spike glycoprotein
+WARNING: This sequence contains 1 non-ACTG character(s) in small secreted glycoprotein
+WARNING: This sequence contains 1 non-ACTG character(s) in second secreted glycoprotein
+
+
@@ -674,0 +769 @@
+
@@ -683,2 +778 @@
-6039	6923	misc_feature
-6923	8069
+6039	8069	misc_feature
@@ -687 +781 @@
-			note	similar to small secreted glycoprotein
+			product	similar to small secreted glycoprotein
@@ -706,2 +800,4 @@
-ERROR: CDS Has Stop Codon: Contains unexpected stop codon in spike glycoprotein; in-frame stop codon exists 5' of stop position predicted by homology to reference [homology search predicted 6923..8069 exon 2 of 2 revised to 6923..7048 (stop shifted 1021 nt)]
-ERROR: Mutation at End: Expected stop codon could not be identified on small secreted glycoprotein; first in-frame stop codon exists 3' of stop position predicted by homology to reference [homology search predicted 6039..7134 revised to 6039..8069 (stop shifted 935 nt)]
+ERROR:  Insertion of Nucleotides: Contains internal insertion of 1 nucleotide(s) in spike glycoprotein
+ERROR:  Insertion of Nucleotides: Contains internal insertion of 1 nucleotide(s) in small secreted glycoprotein
+May include:
+ERROR: Mutation at End: Expected stop codon could not be identified on small secreted glycoprotein
